File: Tests/com/hps/interpreter/LexerTest.java
package com.hps.interpreter;

import org.junit.jupiter.api.Test;
import java.util.ArrayList;
import static org.junit.jupiter.api.Assertions.*;

class LexerTest
{
	@Test
	public void testNextTokenSimple()
	{
		String input = "{}(),;=+";

		Lexer lexer = new Lexer(input);

		ArrayList<Token> expectedTokenList = new ArrayList<Token>();

		expectedTokenList.add(new Token(TokenType.LBRACE, "{"));
		expectedTokenList.add(new Token(TokenType.RBRACE, "}"));
		expectedTokenList.add(new Token(TokenType.LPAREN, "("));
		expectedTokenList.add(new Token(TokenType.RPAREN, ")"));
		expectedTokenList.add(new Token(TokenType.COMMA, ","));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.ASSIGN, "="));
		expectedTokenList.add(new Token(TokenType.PLUS, "+"));
		expectedTokenList.add(new Token(TokenType.EOF, ""));

		for (Token expectedToken : expectedTokenList)
		{
			assertEquals(expectedToken.tokenType, lexer.nextToken().tokenType);
		}
	}

	@Test
	public void testNextToken()
	{
		String input = "let five = 5;\n" +
				"let ten = 10;\n" +
				"\n" +
				"let add = fn(x, y) {\n" +
				"  x + y;\n" +
				"};\n" +
				"\n" +
				"let result = add(five, ten);\n" +
				"!-/*5;\n" +
				"5 < 10 > 5;\n" +
				"\n" +
				"if (5 < 10) {\n" +
				"\treturn true;\n" +
				"} else {\n" +
				"\treturn false;\n" +
				"}\n" +
				"\n" +
				"10 == 10;\n" +
				"10 != 9;";

		Lexer lexer = new Lexer(input);

		ArrayList<Token> expectedTokenList = new ArrayList<Token>();

		expectedTokenList.add(new Token(TokenType.LET, "let"));
		expectedTokenList.add(new Token(TokenType.IDENT, "five"));
		expectedTokenList.add(new Token(TokenType.ASSIGN, "="));
		expectedTokenList.add(new Token(TokenType.INT, "5"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.LET, "let"));
		expectedTokenList.add(new Token(TokenType.IDENT, "ten"));
		expectedTokenList.add(new Token(TokenType.ASSIGN, "="));
		expectedTokenList.add(new Token(TokenType.INT, "10"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.LET, "let"));
		expectedTokenList.add(new Token(TokenType.IDENT, "add"));
		expectedTokenList.add(new Token(TokenType.ASSIGN, "="));
		expectedTokenList.add(new Token(TokenType.FUNCTION, "fn"));
		expectedTokenList.add(new Token(TokenType.LPAREN, "("));
		expectedTokenList.add(new Token(TokenType.IDENT, "x"));
		expectedTokenList.add(new Token(TokenType.COMMA, ","));
		expectedTokenList.add(new Token(TokenType.IDENT, "y"));
		expectedTokenList.add(new Token(TokenType.RPAREN, ")"));
		expectedTokenList.add(new Token(TokenType.LBRACE, "{"));
		expectedTokenList.add(new Token(TokenType.IDENT, "x"));
		expectedTokenList.add(new Token(TokenType.PLUS, "+"));
		expectedTokenList.add(new Token(TokenType.IDENT, "y"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.RBRACE, "}"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.LET, "let"));
		expectedTokenList.add(new Token(TokenType.IDENT, "result"));
		expectedTokenList.add(new Token(TokenType.ASSIGN, "="));
		expectedTokenList.add(new Token(TokenType.IDENT, "add"));
		expectedTokenList.add(new Token(TokenType.LPAREN, "("));
		expectedTokenList.add(new Token(TokenType.IDENT, "five"));
		expectedTokenList.add(new Token(TokenType.COMMA, ","));
		expectedTokenList.add(new Token(TokenType.IDENT, "ten"));
		expectedTokenList.add(new Token(TokenType.RPAREN, ")"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.BANG, "!"));
		expectedTokenList.add(new Token(TokenType.MINUS, "-"));
		expectedTokenList.add(new Token(TokenType.SLASH, "/"));
		expectedTokenList.add(new Token(TokenType.ASTERISK, "*"));
		expectedTokenList.add(new Token(TokenType.INT, "5"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.INT, "5"));
		expectedTokenList.add(new Token(TokenType.LT, "<"));
		expectedTokenList.add(new Token(TokenType.INT, "10"));
		expectedTokenList.add(new Token(TokenType.GT, ">"));
		expectedTokenList.add(new Token(TokenType.INT, "5"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.IF, "if"));
		expectedTokenList.add(new Token(TokenType.LPAREN, "("));
		expectedTokenList.add(new Token(TokenType.INT, "5"));
		expectedTokenList.add(new Token(TokenType.LT, "<"));
		expectedTokenList.add(new Token(TokenType.INT, "10"));
		expectedTokenList.add(new Token(TokenType.RPAREN, "}"));
		expectedTokenList.add(new Token(TokenType.LBRACE, "{"));
		expectedTokenList.add(new Token(TokenType.RETURN, "return"));
		expectedTokenList.add(new Token(TokenType.TRUE, "true"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.RBRACE, "}"));
		expectedTokenList.add(new Token(TokenType.ELSE, "else"));
		expectedTokenList.add(new Token(TokenType.LBRACE, "{"));
		expectedTokenList.add(new Token(TokenType.RETURN, "return"));
		expectedTokenList.add(new Token(TokenType.FALSE, "false"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.RBRACE, "}"));
		expectedTokenList.add(new Token(TokenType.INT, "10"));
		expectedTokenList.add(new Token(TokenType.EQ, "=="));
		expectedTokenList.add(new Token(TokenType.INT, "10"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.INT, "10"));
		expectedTokenList.add(new Token(TokenType.NOT_EQ, "!="));
		expectedTokenList.add(new Token(TokenType.INT, "9"));
		expectedTokenList.add(new Token(TokenType.SEMICOLON, ";"));
		expectedTokenList.add(new Token(TokenType.EOF, ""));

		for (Token expectedToken : expectedTokenList)
		{
			assertEquals(expectedToken.tokenType, lexer.nextToken().tokenType);
		}
	}
}

File: interpreter.tests/LexerTest.cs
using System;
using Xunit;
using System.Collections.Generic;

using interpreter.lexer;

namespace interpreter.tests
{
    public class LexerTest
    {
        [Fact]
        public void GivenNextToken_WhenProvidedWithAValidSetOfCharacters_ThenTheCorrectTokensAreReturned()
        {
            // Arrange
            var input = "=+(){},;";
            var lexer = new Lexer(input);
            var actualTokens = new List<Token>();
            var expectedTokens = new List<Token>
            {
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.PLUS, "+"),
                new Token(TokenType.LPAREN, "("),
                new Token(TokenType.RPAREN, ")"),
                new Token(TokenType.LBRACE, "{"),
                new Token(TokenType.RBRACE, "}"),
                new Token(TokenType.COMMA, ","),
                new Token(TokenType.SEMICOLON, ";"),
            };

            // Act
            foreach (var token in expectedTokens)
            {
                actualTokens.Add(lexer.NextToken());
            }

            // Assert
            Assert.Equal(expectedTokens, actualTokens);
        }

        [Fact]
        public void GivenNextToken_WhenProvidedWithASimpleProgram_TheTheCorrectTokensAreReturned()
        {
            // Arrange
            var input = @"let five = 5;
let ten = 10;

let add = fn(x, y){
x + y;
};

let result = add(five, ten);

!-/*5;
5 < 10 > 5;

if (5 < 10){
return true;
} else {
return false;
}

10 == 10;
10 != 9;";
            var lexer = new Lexer(input);
            var actualTokens = new List<Token>();
            var expectedTokens = new List<Token>
            {
                new Token(TokenType.LET, "let"),
                new Token(TokenType.IDENT, "five"),
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.INT, "5"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.LET, "let"),
                new Token(TokenType.IDENT, "ten"),
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.INT, "10"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.LET, "let"),
                new Token(TokenType.IDENT, "add"),
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.FUNCTION, "fn"),
                new Token(TokenType.LPAREN, "("),
                new Token(TokenType.IDENT, "x"),
                new Token(TokenType.COMMA, ","),
                new Token(TokenType.IDENT, "y"),
                new Token(TokenType.RPAREN, ")"),
                new Token(TokenType.LBRACE, "{"),
                new Token(TokenType.IDENT, "x"),
                new Token(TokenType.PLUS, "+"),
                new Token(TokenType.IDENT, "y"),
                new Token(TokenType.SEMICOLON, ";"),
                new Token(TokenType.RBRACE, "}"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.LET, "let"),
                new Token(TokenType.IDENT, "result"),
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.IDENT, "add"),
                new Token(TokenType.LPAREN, "("),
                new Token(TokenType.IDENT, "five"),
                new Token(TokenType.COMMA, ","),
                new Token(TokenType.IDENT, "ten"),
                new Token(TokenType.RPAREN, ")"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.BANG, "!"),
                new Token(TokenType.MINUS, "-"),
                new Token(TokenType.SLASH, "/"),
                new Token(TokenType.ASTERISK, "*"),
                new Token(TokenType.INT, "5"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.INT, "5"),
                new Token(TokenType.LT, "<"),
                new Token(TokenType.INT, "10"),
                new Token(TokenType.RT, ">"),
                new Token(TokenType.INT, "5"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.IF, "if"),
                new Token(TokenType.LPAREN, "("),
                new Token(TokenType.INT, "5"),
                new Token(TokenType.LT, "<"),
                new Token(TokenType.INT, "10"),
                new Token(TokenType.RPAREN, ")"),
                new Token(TokenType.LBRACE, "{"),
                new Token(TokenType.RETURN, "return"),
                new Token(TokenType.TRUE, "true"),
                new Token(TokenType.SEMICOLON, ";"),
                new Token(TokenType.RBRACE, "}"),
                new Token(TokenType.ELSE, "else"),
                new Token(TokenType.LBRACE, "{"),
                new Token(TokenType.RETURN, "return"),
                new Token(TokenType.FALSE, "false"),
                new Token(TokenType.SEMICOLON, ";"),
                new Token(TokenType.RBRACE, "}"),

                new Token(TokenType.INT, "10"),
                new Token(TokenType.EQ, "=="),
                new Token(TokenType.INT, "10"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.INT, "10"),
                new Token(TokenType.NOT_EQ, "!="),
                new Token(TokenType.INT, "9"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.EOF, "")
            };

            // Act
            foreach (var token in expectedTokens)
            {
                actualTokens.Add(lexer.NextToken());
            }

            for (int i = 0; i < actualTokens.Count; i++)
            {
                if (!expectedTokens[i].Equals(actualTokens[i]))
                {
                    Console.WriteLine("agh");
                }
            }

        // Assert
            Assert.Equal(expectedTokens, actualTokens);
        }

    }
}
